core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -75,10 +93,26 @@ public static TableMetadata newTableMetadata(TableOperations ops,
     }
     PartitionSpec freshSpec = specBuilder.build();
 
+    // rebuild the sort order using the new column ids
+    SortOrder.Builder sortOrderBuilder = SortOrder.builderFor(freshSchema)
+        .withOrderId(INITIAL_SORT_ORDER_ID);
+    for (SortField field : sortOrder.fields()) {
+      int[] sourceIds = field.sourceIds();
+      int[] freshSourceIds = new int[sourceIds.length];
+      for (int i = 0; i < sourceIds.length; i++) {
+        // look up the name of the source field in the old schema to get the new schema's id
+        String sourceName = schema.findColumnName(sourceIds[i]);
+        freshSourceIds[i] = freshSchema.findField(sourceName).fieldId();
+      }
+      sortOrderBuilder.add(field.name(), field.direction(), field.nullOrder(), field.transform(), freshSourceIds);
+    }
+    SortOrder freshSortOrder = sortOrderBuilder.build();
+
     return new TableMetadata(ops, null, UUID.randomUUID().toString(), location,
         System.currentTimeMillis(),
-        lastColumnId.get(), freshSchema, INITIAL_SPEC_ID, ImmutableList.of(freshSpec),
-        ImmutableMap.copyOf(properties), -1, ImmutableList.of(), ImmutableList.of());
+        lastColumnId.get(), freshSchema, INITIAL_SPEC_ID, ImmutableList.of(freshSpec), INITIAL_SORT_ORDER_ID,
+        ImmutableList.of(freshSortOrder), ImmutableMap.copyOf(properties), -1,
+        ImmutableList.of(), ImmutableList.of());
```

Comment:
Nit: probably `Collections.emptyList()` fits best if immutability is the main requirement here, JDK over library.